code minings drawn more consistently always via drawAsLeftOf1stCharacter

tobiasmelcher opened the pull request:
this improves the selection behavior for the scenario where the selection end offset contains a LineContentAnnotation. Selection is no longer extended with the code mining.

The selection behavior before this change always included the code mining text which is located at the end selection offset. The following screenshot shows the behavior. By including the last character "o" of "echo" in the selection, the selection is extended with the code mining text next to the "o" character.
![code_mining_draw_right_to_the_previous_character](https://github.com/user-attachments/assets/ca310b43-aa15-43ef-adf5-50ef65b67ab6)

This is no longer the case with this change. The code mining text is no longer added to the selection when "echo" is selected.
![code_mining_draw_at_current_character](https://github.com/user-attachments/assets/78fcee62-1c0d-4d58-9425-fbfd5ca1eebe)

Is the current behavior of drawing the code mining texts intended? Is this behavior needed for a feature (which I am not aware of)?

Let me try to explain why I request this change.
In the ABAP Development Tools in Eclipse we use the code minings to show inline keyword completions while the user is typing a keyword. If a matching keyword is available, it is shown as code mining and user can take over the text by pressing the tab key. Here a screencast showing the scenario when the user is typing the keyword "definition". 

https://github.com/user-attachments/assets/8770b291-08e6-4cd2-ba83-bf1a67d6c93c

It works quite well. Users don't like the way how the cursor always jumps to the end of the code mining showing the rest of the keyword "definition". The cursor jumps to the end of the code mining because method drawAsRightOfPreviousCharacter is used to render the code mining. 

The cursor does not jump to the end of the code mining with this pull request by using method drawAsLeftOf1stCharacter to render the code mining.

https://github.com/user-attachments/assets/111a41bf-1c9d-4c46-9e89-17965c05a1d4

The question now is if we would break any existing functionality regarding rendering of code minings by always using method drawAsLeftOf1stCharacter?
An alternative approach would be to provide an API with which the user can decide how to draw the code mining. What would be your preferred solution? Could you please give any guidance?

Thanks a lot,
Tobias

BeckerWdf commented:
"pr-head" says:
` Failed to execute goal org.eclipse.tycho.extras:tycho-p2-extras-plugin:4.0.8:compare-version-with-baselines (compare-attached-artifacts-with-release) on project org.eclipse.jface.text: Only qualifier changed for (org.eclipse.jface.text/3.25.200.v20240905-1304). Expected to have bigger x.y.z than what is available in baseline (3.25.200.v20240816-1151)`

Pls add a second mitt that increases the version of org.eclipse.jface-text to "3.25.300". This should fix this issue.

tobias-melcher committed:
increase version numbers for next release

BeckerWdf commented:
There seems to be an infrastructure issue:

```
06:14:56.854 [ERROR] Error executing Maven.
06:14:56.855 [ERROR] Extension org.eclipse.tycho:tycho-build:4.0.6 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.eclipse.tycho:tycho-build:jar:4.0.6
06:14:56.856 [ERROR] Caused by: Failed to read artifact descriptor for org.eclipse.tycho:tycho-build:jar:4.0.6
06:14:56.856 [ERROR] Caused by: Failed to read artifact descriptor for org.eclipse.tycho:tycho-build:jar:4.0.6
06:14:56.856 [ERROR] Caused by: The following artifacts could not be resolved: org.eclipse.tycho:tycho-build:pom:4.0.6 (absent): Could not transfer artifact org.eclipse.tycho:tycho-build:pom:4.0.6 from/to eclipse.maven.central.mirror (https://repo.eclipse.org/content/repositories/maven_central/): status code: 403, reason phrase: Forbidden (403)
06:14:56.856 [ERROR] Caused by: Could not transfer artifact org.eclipse.tycho:tycho-build:pom:4.0.6 from/to eclipse.maven.central.mirror (https://repo.eclipse.org/content/repositories/maven_central/): status code: 403, reason phrase: Forbidden (403)
06:14:56.857 [ERROR] Caused by: status code: 403, reason phrase: Forbidden (403)
```

BeckerWdf commented:
@mickaelistria: What are your thoughts on this PR?

BeckerWdf commented:
org.eclipse.jface.text.tests also needs a version bump:

`Failed to execute goal org.eclipse.tycho.extras:tycho-p2-extras-plugin:4.0.8:compare-version-with-baselines (compare-attached-artifacts-with-release) on project org.eclipse.jface.text.tests: Only qualifier changed for (org.eclipse.jface.text.tests/3.13.600.v20240913-0631). Expected to have bigger x.y.z than what is available in baseline (3.13.600.v20240816-1151) -> [Help 1]`

Can you pls. ammend your version change commit?

BeckerWdf commented:
Some of the code mining tests are failing:

org.eclipse.jface.text.tests.codemining.CodeMiningTest.testCodeMiningOnEmptyLine
org.eclipse.jface.text.tests.codemining.CodeMiningTest.testCodeMiningAtEndOfLine

BeckerWdf commented:
> org.eclipse.jface.text.tests also needs a version bump:
> 
> `Failed to execute goal org.eclipse.tycho.extras:tycho-p2-extras-plugin:4.0.8:compare-version-with-baselines (compare-attached-artifacts-with-release) on project org.eclipse.jface.text.tests: Only qualifier changed for (org.eclipse.jface.text.tests/3.13.600.v20240913-0631). Expected to have bigger x.y.z than what is available in baseline (3.13.600.v20240816-1151) -> [Help 1]`
> 
> Can you pls. ammend your version change commit?

Thanks

mickaelistria commented:
I didn't try this PR yet.
But the history here is that the main use-case for inline code minings in the Eclipse SDK are currently method parameter annotation. With those, user want  for example `System.arraycopy((src: array1, srcPos: |0, dest: array2, destPos: 0, n)`. If we change it the other way round, then we break this case, the cursor before `0` would be ~80 pixels to the left. Both this legacy use-case and the one you're sharing here seem antagonist. We most likely need some extra flag on codeMinings somewhere to declare whether it's a prefix or a suffix.

tobiasmelcher commented:
Thanks a lot for the feedback @mickaelistria. I fully agree. I just tested the java method parameter code minings with this change and selection behavior looks ugly. 

Without this change, one can select the method parameter value 0 and the code mining text is not included.
![java_method_param_before](https://github.com/user-attachments/assets/330edc29-815a-44b5-b9f9-c0c07857e712)

With this change, this is no longer possible. Selecting 0 always includes the code mining which does not feel right.
![java_method_param_after](https://github.com/user-attachments/assets/a6a6a81e-01f2-4d94-9fe9-f8e46f0554e2)

I will make a proposal with an API enhancement so that the selection behavior of the code minings can be configured. For sure in a compatible way.

BeckerWdf commented:
> I will make a proposal with an API enhancement so that the selection behavior of the code minings can be configured. For sure in a compatible way.

Yes. That's a good idea.

tobias-melcher committed:
additional constructor param afterPosition in LineContentCodeMining

which allows to render a code mining where the cursor selection does not
include the code mining at the given source position.

mickaelistria commented on bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/codemining/SuffixLineContentCodeMining.java:
```diff
@@ -0,0 +1,45 @@
+ * Line content code mining which is drawn right to the given position. Cursor and selection
+ * handling are not including the code mining.
+ */
+public abstract class SuffixLineContentCodeMining extends LineContentCodeMining {
```
I would rather imagine a new flag on LineContentCodeMning than a whole new method here.
IMO, a new constructor `LineContentCodeMining(Position, boolean afterPosition /* default being false in other constructors that don't have this arg */, provider)` would be simpler to both consume and maintain.

tobiasmelcher replied:
Thanks a lot for the fast reply. I changed the API as requested and added a further constructor in LineContentCodeMining. I have put "boolean afterPosition" as last contructor argument for compatibility reasons. Do you think it is better to place it as second argument?

mickaelistria replied:
For compatibility, you only need to keep the current constructors. You're free to design your new ones freely.
So yes, I think it's better to place it as 2nd argument if it's not conflicting with any other existing constructor.

tobiasmelcher replied:
ok, I added it as 2nd argument. Is it because first argument is "position" and next argument "afterPosition" is then influencing the position?

mickaelistria replied:
Yes, the position where the codemining is rendered is a function of those 2 arguments. Grouping them together makes things faster to understand.

tobiasmelcher replied:
ok, understood. Makes absolutely sense.

mickaelistria reviewed (commented):
I have not tested it, but the code looks already good.
Just a comment inline discussing whether a new CodeMining type is useful here.

BeckerWdf commented on examples/org.eclipse.jface.text.examples/META-INF/MANIFEST.MF:
```diff
@@ -2,7 +2,7 @@ Manifest-Version: 1.0
 Bundle-Name: CodeMinig Examples
 Bundle-SymbolicName: org.eclipse.jface.text.examples
-Bundle-Version: 1.2.0.qualifier
+Bundle-Version: 1.2.1.qualifier
```
i think this has to be. 1.2.100

tobiasmelcher replied:
thanks, should be corrected with the last upload

mickaelistria commented on bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/codemining/LineContentCodeMining.java:
```diff
@@ -44,7 +62,25 @@ public LineContentCodeMining(Position position, ICodeMiningProvider provider) {
+	 * @param afterPosition if true code mining is treated as suffix code mining where cursor and
+	 *            selection is not including the mining
+	 */
+	public LineContentCodeMining(Position position, boolean afterPosition, ICodeMiningProvider provider, Consumer<MouseEvent> action) {
```
Since this is new API, this will likely require some `@since 3.26` and a version increase in the MANIFEST.

mickaelistria commented on bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/source/inlined/LineContentAnnotation.java:
```diff
@@ -158,7 +158,7 @@ StyleRange updateStyle(StyleRange style, FontMetrics fontMetrics, ITextViewer vi
 	}
 
-	static boolean drawRightToPreviousChar(int widgetOffset, StyledText textWidget) {
+	protected boolean drawRightToPreviousChar(int widgetOffset, StyledText textWidget) {
```
Since this is new API, this will likely require some `@since 3.26` and a version increase in the MANIFEST.
Or better if you can find a way to not make this method API.

tobiasmelcher replied:
thanks a lot for the help. I tried to fix the API tools messages with 80b1736. I added an additional argument to package method LineContentAnnotation#updateStyle . Is this allowed? If not, I would need to introduce method "protected boolean drawRightToPreviousChar" as protected instance method.

mickaelistria reviewed (commented):
Some notes about probable complaints from API tools about new API methods.

mickaelistria commented:
>  Is this allowed?

Yes, it is allowed because this method is package visible only (not API then).

mickaelistria commented on bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/codemining/LineContentCodeMining.java:
```diff
@@ -26,6 +26,14 @@
+	 *
+	 * @since 3.26
+	 */
+	protected boolean afterPosition= false;
```
As there is a getter, couldn't this one be `private final` too?

tobiasmelcher replied:
I fully agree, thanks for the feedback.

mickaelistria reviewed (approved):
Code looks very good and the use-case is pretty interesting. Let's merge it as soon as CI reports a successful build on Jenkins.

BeckerWdf commented:
failed test are unrelated to this PR

mickaelistria commented:
Thank you for this good code and for making the IDE capable of achieving 1 new use-case!
